plugin/db/util/driverutil.go

```diff
@@ -132,151 +132,174 @@ type MigrationExecutor interface {
 // Returns the created migraiton history id and the updated schema on success.
 func ExecuteMigration(ctx context.Context, l *zap.Logger, executor MigrationExecutor, m *db.MigrationInfo, statement string) (migrationHistoryID int64, updatedSchema string, resErr error) {
 	var prevSchemaBuf bytes.Buffer
-	// Don't record schema if the database hasn't exist yet.
-	if !m.CreateDatabase {
+	if m.CreateDatabase {
+		// Don't record schema if the database hasn't exist yet.
+	} else if m.Type == db.Baseline {
+		// Record the updated schema of the previous migration if exists.
+		prevMigrationHistoryFind := db.MigrationHistoryFind{
+			Database: &m.Database,
+			Limit:    new(int),
+		}
+		*prevMigrationHistoryFind.Limit = 1
+		migrationHistoryList, err := executor.FindMigrationHistoryList(ctx, &prevMigrationHistoryFind)
+		if err != nil {
+			l.Error("Failed to find migration history", zap.Error(err))
+			return -1, "", err
+		}
+		if len(migrationHistoryList) > 0 {
+			prevSchemaBuf.WriteString(migrationHistoryList[0].Schema)
+		}
+	} else {
+		// Otherwise, dump the schema before the migration.
 		if err := executor.Dump(ctx, m.Database, &prevSchemaBuf, true /*schemaOnly*/); err != nil {
 			return -1, "", formatError(err)
 		}
 	}
 
-	sqldb, err := executor.GetDbConnection(ctx, bytebaseDatabase)
+	// Phase 1 - Precheck before executing migration
+	// Phase 2 - Record migration history as PENDING
+	insertedID, err := preMigrate(ctx, executor, m, prevSchemaBuf.String(), statement)
 	if err != nil {
 		return -1, "", err
 	}
+
+	// If we have already started migration, there will be a PENDING migration record. Upon returning,
+	// we will update that record to DONE or FAILED depending on whether error occurs.
+	// Returned values are passed by reference so that postMigrate() can get the final values.
+	defer postMigrate(ctx, l, executor, time.Now().UnixNano(), &migrationHistoryID, &updatedSchema, &resErr)
```

Comment:
I feel like we're a little bit overusing defer() where defer should only be doing very simple things. Also, the name is called postMigrate but we're actually doing the migrate execution within it.